aiohttp client doesn't follow redirects of mocked requests

Hello,

I'm using aioresponses along with aiohttp client and I'm seeing a problem regarding `aiohttp.client.ClientSession` not following redirects of a mocked request. Here are two tests that reproduces the problem:

```python
from aiohttp.client import ClientSession
from aioresponses import aioresponses
from asynctest import TestCase


class AIOResponseRedirectTest(TestCase):
    async def test_aiohttp_follows_mocked_redirect(self):
        with aioresponses() as rsps:
            rsps.get(
                "http://10.1.1.1:8080/redirect",
                status=307,
                headers={"Location": "https://httpbin.org"},
            )
            client = ClientSession()
            response = await client.get(
                "http://10.1.1.1:8080/redirect", allow_redirects=True
            )
            self.assertEqual(200, response.status)

    async def test_aiohttp_follows_passthrough_redirect(self):
        with aioresponses(passthrough=["https://httpbin.org"]) as rsps:
            rsps.get(
                "http://10.1.1.1:8080/redirect",
                status=307,
                headers={"Location": "https://httpbin.org"},
            )
            client = ClientSession()
            response = await client.get(
                "https://httpbin.org/redirect-to?status_code=307&url=https://httpbin.org",
                allow_redirects=True,
            )
            self.assertEqual(200, response.status)

```

What I expected to see is both tests passing, but when I run this testclass, that's what I see:

```
pipenv run py.test -v tests/test_aiohttp_redirect_bug.py
Loading .env environment variables…
===================================================================================== test session starts =====================================================================================
platform linux -- Python 3.6.8, pytest-3.0.1, py-1.8.0, pluggy-0.3.1 -- /home/daltonmatos/.local/share/virtualenvs/asgard-api-YwspFh3W/bin/python3
cachedir: .cache
rootdir: /home/daltonmatos/src/asgard-api, inifile: 
plugins: cov-2.4.0
collected 2 items 

tests/test_aiohttp_redirect_bug.py::AIOResponseRedirectTest::test_aiohttp_follows_mocked_redirect FAILED
tests/test_aiohttp_redirect_bug.py::AIOResponseRedirectTest::test_aiohttp_follows_passthrough_redirect PASSED

========================================================================================== FAILURES ===========================================================================================
________________________________________________________________ AIOResponseRedirectTest.test_aiohttp_follows_mocked_redirect _________________________________________________________________

self = <tests.test_aiohttp_redirect_bug.AIOResponseRedirectTest testMethod=test_aiohttp_follows_mocked_redirect>

    async def test_aiohttp_follows_mocked_redirect(self):
        with aioresponses() as rsps:
            rsps.get(
                "http://10.1.1.1:8080/redirect",
                status=307,
                headers={"Location": "https://httpbin.org"},
            )
            client = ClientSession()
            response = await client.get(
                "http://10.1.1.1:8080/redirect", allow_redirects=True
            )
>           self.assertEqual(200, response.status)
E           AssertionError: 200 != 307

tests/test_aiohttp_redirect_bug.py:18: AssertionError
============================================================================= 1 failed, 1 passed in 1.18 seconds ==============================================================================

```

Shouldn't the failed test return a `status=200` with the `history` attribute containing the redirected request? Is there anything that I'm doing wrong?

The versions are:

```
$ pipenv run pip freeze | grep -E "aiohttp|aioresp"
Loading .env environment variables…
aiohttp==3.4.4
aiohttp-cors==0.7.0
aioresponses==0.6.0
```

Python is `3.6.8`:

```
$ pipenv run python -V
Loading .env environment variables…
Python 3.6.8
```

Thanks,

--- a/aioresponses/core.py
+++ b/aioresponses/core.py
@@ -299,19 +299,39 @@ class aioresponses(object):
         ))
 
     async def match(
-            self, method: str, url: URL, **kwargs: Dict
+        self, method: str, url: URL,
+        allow_redirects: bool = True, **kwargs: Dict
     ) -> Optional['ClientResponse']:
-        for i, matcher in enumerate(self._matches):
-            if matcher.match(method, url):
-                response = await matcher.build_response(url, **kwargs)
-                break
-        else:
-            return None
+        history = []
+        while True:
+            for i, matcher in enumerate(self._matches):
+                if matcher.match(method, url):
+                    response = await matcher.build_response(
+                        url, allow_redirects=allow_redirects, **kwargs
+                    )
+                    break
+            else:
+                if history:
+                    raise Exception('Redirect leads to an unmocked url')
+                return None
 
-        if matcher.repeat is False:
-            del self._matches[i]
-        if isinstance(response, Exception):
-            raise response
+            if matcher.repeat is False:
+                del self._matches[i]
+            if isinstance(response, Exception):
+                raise response
+
+            if response.status in (
+                    301, 302, 303, 307, 308) and allow_redirects:
+                if hdrs.LOCATION not in response.headers:
+                    break
+                history.append(response)
+                url = URL(response.headers[hdrs.LOCATION])
+                continue
+            else:
+                break
+
+        response._history = tuple(history)
+
         return response
 
     async def _request_mock(self, orig_self: ClientSession,

--- a/aioresponses/core.py
+++ b/aioresponses/core.py
@@ -311,8 +311,6 @@ class aioresponses(object):
                     )
                     break
             else:
-                if history:
-                    raise Exception('Redirect leads to an unmocked url')
                 return None
 
             if matcher.repeat is False:

--- a/tests/test_aioresponses.py
+++ b/tests/test_aioresponses.py
@@ -407,3 +407,56 @@ class AIOResponsesRaiseForStatusSessionTestCase(TestCase):
                                                raise_for_status=False)
 
         self.assertEqual(response.status, 400)
+
+
+class AIOResponseRedirectTest(TestCase):
+    @asyncio.coroutine
+    def setUp(self):
+        self.url = "http://10.1.1.1:8080/redirect"
+        self.session = ClientSession()
+        super().setUp()
+
+    @asyncio.coroutine
+    def tearDown(self):
+        close_result = self.session.close()
+        if close_result is not None:
+            yield from close_result
+        super().tearDown()
+
+    @aioresponses()
+    @asyncio.coroutine
+    def test_redirect_followed(self, rsps):
+        rsps.get(
+            self.url,
+            status=307,
+            headers={"Location": "https://httpbin.org"},
+        )
+        rsps.get("https://httpbin.org")
+        response = yield from self.session.get(
+            self.url, allow_redirects=True
+        )
+        self.assertEqual(response.status, 200)
+        self.assertEqual(str(response.url), "https://httpbin.org")
+        self.assertEqual(len(response.history), 1)
+        self.assertEqual(str(response.history[0].url), self.url)
+
+    @aioresponses()
+    @asyncio.coroutine
+    def test_redirect_missing_mocked_match(self, rsps):
+        rsps.get(
+            self.url,
+            status=307,
+            headers={"Location": "https://httpbin.org"},
+        )
+        with self.assertRaises(Exception) as cm:
+            response = yield from self.session.get(
+                self.url, allow_redirects=True
+            )
+        self.assertEqual(str(cm.exception), 'Redirect leads to an unmocked url')
+
+    @aioresponses()
+    @asyncio.coroutine
+    def test_redirect_missing_location_header(self, rsps):
+        rsps.get(self.url, status=307)
+        response = yield from self.session.get(self.url, allow_redirects=True)
+        self.assertEqual(str(response.url), self.url)

--- a/tests/test_aioresponses.py
+++ b/tests/test_aioresponses.py
@@ -448,11 +448,11 @@ class AIOResponseRedirectTest(TestCase):
             status=307,
             headers={"Location": "https://httpbin.org"},
         )
-        with self.assertRaises(Exception) as cm:
+        with self.assertRaises(ClientConnectionError) as cm:
             response = yield from self.session.get(
                 self.url, allow_redirects=True
             )
-        self.assertEqual(str(cm.exception), 'Redirect leads to an unmocked url')
+        self.assertEqual(str(cm.exception), 'Connection refused: GET http://10.1.1.1:8080/redirect')
 
     @aioresponses()
     @asyncio.coroutine

